packages/hardhat-core/src/internal/hardhat-network/provider/MiningTimer.ts

```diff
@@ -0,0 +1,76 @@
+enum MiningTimerState {
+  STOP,
+  RUNNING,
+}
+
+// tslint:disable only-hardhat-error
+
+export class MiningTimer {
+  private _state = MiningTimerState.STOP;
+  private _timeout: any = null;
+
+  constructor(
+    private _blockTime: number,
+    private readonly _mineFunction: () => Promise<any>
+  ) {
+    if (_blockTime <= 0) {
+      throw new Error(
+        "Block time passed to the constructor must be greater than 0 ms"
+      );
+    }
+  }
+
+  public getBlockTime(): number {
+    return this._blockTime;
+  }
+
+  public setBlockTime(blockTime: number): void {
+    if (blockTime <= 0) {
+      throw new Error("New block time must be greater than 0 ms");
+    }
+
+    if (blockTime === this._blockTime) {
+      return;
+    }
+
+    this._blockTime = blockTime;
+
+    if (this._state === MiningTimerState.RUNNING) {
+      this.stop();
+      this.start();
+    }
+  }
+
+  public start(): void {
+    if (this._state === MiningTimerState.RUNNING) {
+      return;
+    }
+
+    this._state = MiningTimerState.RUNNING;
+    this._timeout = setTimeout(
+      () => this._loop().catch(console.error),
```

Comment:
I did 2, seems more reasonable to me. I also logged the error, since we probably want to know if it happens.

Commit: f3b69cc11b44691bb272a6f7eab8190aaefa1495